lib/config.cpp

```diff
@@ -732,26 +747,21 @@ std::map<std::string, std::vector<Fulfillment>> Config::get_fulfillments(const s
 
     std::map<std::string, std::vector<Fulfillment>> res;
 
-    const std::string module_name = get_module_name(module_id);
-    for (const std::string& req_id : Config::keys(this->manifests.at(module_name).at("requires"))) {
-        std::vector<Fulfillment> fulfillments;
-        const json resolved_req = this->resolve_requirement(module_id, req_id);
-        if (!resolved_req.is_array()) {
-            const auto resolved_module_id = resolved_req.at("module_id");
-            const auto resolved_impl_id = resolved_req.at("implementation_id");
-            const auto mapping = this->get_3_tier_model_mapping(resolved_module_id, resolved_impl_id);
-            Requirement req(req_id, 0);
-            fulfillments.push_back({resolved_module_id, resolved_impl_id, req, mapping});
-        } else {
-            for (int i = 0; i < resolved_req.size(); i++) {
-                const auto resolved_module_id = resolved_req.at(i).at("module_id");
-                const auto resolved_impl_id = resolved_req.at(i).at("implementation_id");
-                const auto mapping = this->get_3_tier_model_mapping(resolved_module_id, resolved_impl_id);
-                Requirement req(req_id, i);
-                fulfillments.push_back({resolved_module_id, resolved_impl_id, req, mapping});
-            }
-        }
-        res[req_id] = fulfillments;
+    for (const auto& [requirement, fulfillment] : this->resolve_requirements(module_id)) {
+        res[requirement.id].push_back(fulfillment);
+    }
+
+    return res;
+}
+
+RequirementInitialization Config::get_requirement_initialization(const std::string& module_id) const {
+    BOOST_LOG_FUNCTION();
+
+    RequirementInitialization res;
+
+    for (const auto& [requirement, fulfillment] : this->resolve_requirements(module_id)) {
+        const auto& mapping = this->get_3_tier_model_mapping(fulfillment.module_id, fulfillment.implementation_id);
+        res[requirement.id].push_back({requirement, fulfillment, mapping});
```

Comment:
This is a bit tricky. We implicitly rely on the fact, that requirements are sorted within the map according to their index.  This should be guaranteed, but still is not obvious to the reader.  Having the requirement without it's index, but instead pointing at a vector of `Fulfillments` would probably make the code even simpler.